proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -156,12 +144,500 @@ vector<float, 3> ApplyNeuralMaterial(vector<half, 8> inputVector) {
 
 ## Detailed design
 
-TBD
+The API is implemented using the C++03 style templates supported in HLSL 2021.
+Therefore HLSL 2021 is required to use this API.
+
+Developers are expected to rely on template argument deduction for calls to the
+various functions defined. In this document a higher-level view of how
+developers should view each function is provided, along with a proposed
+implementation.
 
 ### Builtins
 
 TBD
 
+### enum DataType
+
+This enum defines the various data types that can be applied to matrices and
+vectors. The numeric values of these enum entries are picked to match those used
+in the underlying DXIL.
+
+```c++
+enum DataType {
+  DATA_TYPE_SINT16 = 2,           // ComponentType::I16
+  DATA_TYPE_UINT16 = 3,           // ComponentType::U16
+  DATA_TYPE_SINT32 = 4,           // ComponentType::I32
+  DATA_TYPE_UINT32 = 5,           // ComponentType::U32
+  DATA_TYPE_FLOAT16 = 8,          // ComponentType::F16
+  DATA_TYPE_FLOAT32 = 9,          // ComponentType::F32
+  DATA_TYPE_SINT8_T4_PACKED = 17, // ComponentType::PackedS8x32
+  DATA_TYPE_UINT8_T4_PACKED = 18, // ComponentType::PackedU8x32
+  DATA_TYPE_UINT8 = 19,           // ComponentType::U8
+  DATA_TYPE_SINT8 = 20,           // ComponentType::I8
+  DATA_TYPE_E4M3 = 21,            // ComponentType::F8_E4M3
+                                  // (1 sign, 4 exp, 3 mantissa bits)
+  DATA_TYPE_E5M2 = 22,            // ComponentType::F8_E5M2
+                                  // (1 sign, 5 exp, 2 mantissa bits)
+};
+```
+
+See the Type Interpretations section of [0029] for more information.
+
+### enum MatrixLayout
+
+This enum defines the different possible layouts of a matrix in memory.
+
+```c++
+enum MatrixLayout {
+  MATRIX_LAYOUT_ROW_MAJOR = 0,
+  MATRIX_LAYOUT_COLUMN_MAJOR = 1,
+  MATRIX_LAYOUT_MUL_OPTIMAL = 2,
+  MATRIX_LAYOUT_OUTER_PRODUCT_OPTIMAL = 3
+};
+```
+
+See the Matrix Layouts section of [0029] for more information.
+
+### struct MatrixRef / RWMatrixRef
+
+`MatrixRef` and `RWMatrixRef` specify a reference to a matrix in memory,
+including information about the dimensions, layout and numerical format of the
+matrix. Some of this information must be known at compile time - these are
+captured in template parameters - while others can be determined at runtime, and
+are therefore members of the struct.
+
+Since the `ByteAddressBuffer` and `RWByteAddressBuffer` are not convertable to
+each other, use use different named types for the matrix references. A base
+class, `MatrixRefImpl` is used for the common code.
+
+Other functions in the API are implemented in terms of `MatrixRefImpl`. This
+allows the compiler to generate more useful diagnostics if wrong types are used.
```

Comment:
Would it be clearer to say 'unsupported types' instead of 'wrong types'?

Reply:
Wrote a bit more here to try and explain more about the design choice made here.